src/strategies/flex_window.py

```diff
@@ -84,170 +86,321 @@ def step(self, event_list=[]):
             sim_vehicle = deepcopy(vehicle)
-            if self.LOAD_STRAT == 'balanced':
-                # simple case: charge balanced during windows
-                # try to charge with full power
+            # simple case: charge balanced during windows
+            # try to charge with full power
+            cur_time = self.current_time - self.interval
+            for ts_info in timesteps:
+                cur_time += self.interval
+                if cur_time >= sim_vehicle.estimated_time_of_departure:
+                    break
+                if ts_info["window"]:
+                    p = ts_info["power"]
+                    p = util.clamp_power(p, sim_vehicle, cs)
+                    sim_vehicle.battery.load(self.interval, p)
+
+            charged_in_window = sim_vehicle.get_delta_soc() <= 0
+
+            if charged_in_window:
+                # reset sim SoC
+                sim_vehicle.battery.soc = vehicle.battery.soc
+
+            min_power = 0
+            max_power = util.clamp_power(cs.max_power, sim_vehicle, cs)
+            old_soc = sim_vehicle.battery.soc
+            safe = False
+            power_vec = [0] * len(timesteps)
+            while (
+                charged_in_window and not safe
+            ) or max_power - min_power > self.EPS:
+                power = (min_power + max_power) / 2
+                sim_vehicle.battery.soc = old_soc
+
                 cur_time = self.current_time - self.interval
-                for ts_info in timesteps:
+                for ts_idx, ts_info in enumerate(timesteps):
                     cur_time += self.interval
+                    avg_power = 0
                     if cur_time >= sim_vehicle.estimated_time_of_departure:
                         break
-                    if ts_info["window"]:
-                        p = ts_info["power"]
-                        p = util.clamp_power(p, sim_vehicle, cs)
-                        sim_vehicle.battery.load(self.interval, p)
+                    if ts_info["window"] == charged_in_window:
+                        # if ts_info["window"] == False:
+                        #     print("stop")
+                        p = util.clamp_power(power, sim_vehicle, cs)
+                        avg_power = sim_vehicle.battery.load(self.interval, p)[
+                            "avg_power"
+                        ]
+                    elif not charged_in_window and ts_info["window"]:
+                        # charging windows not sufficient, charge max during window
+                        p = util.clamp_power(ts_info["power"], sim_vehicle, cs)
+                        avg_power = sim_vehicle.battery.load(self.interval, p)[
+                            "avg_power"
+                        ]
+
+                    power_vec[ts_idx] = avg_power
+                    safe = sim_vehicle.get_delta_soc() <= 0
+                    if safe:
+                        # todo: why is this?
```

Comment:
looks like it's resetting every timestep to 0 after a safe power level is found to avoid having old results in the power vector from prior unsafe simulations